### PR TITLE
Hscript console, closes #1156

### LINKED ISSUE #1156: Consider partially rewriting the console to use hscript
From what I've seen, hscript is quite powerful, and we would support a subset of haxe code from within the console basically. I think this could be done in a way that doesn't make hscript a dependency. (i.e. when trying to use a command that uses hscript, print an error in the console saying it has to be installed, and cover the relevant code parts in `#if hscript`).

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/2538874-consider-partially-rewriting-the-console-to-use-hscript?utm_campaign=plugin&utm_content=tracker%2F254189&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F254189&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by MSGhero**
The console itself works fine now, but *now* I need some confirmation that these changes are ok. There are some breaking changes now.

### CODE REVIEW

**Comment by larsiusprime** [comment]
neato

**Comment by MSGhero** [comment]
I'm mainly done now, but yeah it's a breaking change. Since hscript deals with objects directly instead of strings, I removed all the parsing code and string parameters from functions.

**Comment by MSGhero** [comment]
Pausing and TAB don't work on neko, but they probably never did. Didn't test html5.

**Comment by MSGhero** [comment]
The travis script and unit tests work fine on my computer, and the build does install hscript. Why can't it find hscript.Expr?

**Comment by Gama11** [comment]
As is, every flixel project that doesn't include hscript in the xml manually won't compile - would be better to have that in the include.xml.

In theory it would be nice to only include it if it's installed (and only have the console be usable in that case), but I'm not sure if there's a way to do that with the project xml format...

**Comment by MSGhero** [comment]
I added it to include.xml, but would it go somewhere besides the `setup` block? Should I wrap all console code in `#if hscript`?

**Comment by Gama11** [comment]
`setup` is only used for `lime setup flixel`. If you actually want to include the haxelib during compilation, you need the regular `<haxelib name="hscript" />` like the `include.xml` does with openfl.

**Comment by MSGhero** [comment]
Ugh ok. I see what you mean now about always including it. I guess it would always be included for `release` builds too even when the debugger flxdef is off.

**Comment by MSGhero** [comment]
The unit tests look fine, but it's still breaking coverage on the `import hscript.Expr`

**Comment by Gama11** [comment]
The unit tests use a .hxml file for flash.

https://github.com/HaxeFlixel/flixel/blob/dev/tests/unit/test.hxml

**Comment by MSGhero** [comment]
I'm never adding an external lib to flixel again.

**Comment by MSGhero** [comment]
Oh hey, dev cpp is working again. Haxe recently fixed the multiple inheritance issue which was preventing flixel-ui from compiling.

**Comment by gamedevsam** [comment]
This change looks great! Has anyone been able to test this functionality?

**Comment by MSGhero** [comment]
Me lol. Do you need a rundown of the changes or are you making sure it works for other people?

**Comment by gamedevsam** [comment]
I was mostly wondering if there was someone in the community that opposes this change or has something to say before I merge it in. @HaxeFlixel/owners any opposition or can I merge this?

**Comment by Beeblerox** [comment]
i haven't checked this pull request, but i don't mind replacing current functionality with hscript as far as it works

File: flixel/system/debug/ConsoleUtil.hx [hunk 1]
```diff
@@ -380,9 +85,20 @@ class ConsoleUtil
 }
 
 /**
- * Data necessary to set a variable via Reflection. Used by the set command / resolveObjectAndVariable().
+ * hscript doesn't use property access by default... have to make our own.
  */
-typedef PathToVariable = {
-	object:Dynamic,
-	variableName:String
+private class Interp extends hscript.Interp
```
Comment (MSGhero) [inline flixel/system/debug/ConsoleUtil.hx]: FlxInterp? lol

**Comment by MSGhero** [comment]
I'm not 100% sure if the Watch tab still works all the time, but the `watch` console command does work.

I'm also not sure why the `Event.RENDER` bit was in there, maybe a workaround for an openfl thing.

**Comment by Gama11** [comment]
The ability to run arbitrary Haxe code from the console is pretty amazing. :)

Also:
>202 additions and 809 deletions

:+1: 

Not too mention, seems like it's not as crashy on neko anymore.

**Comment by MSGhero** [comment]
In the original issue, you said that it'd be cool if we could `#if hscript` to display an error message. Is that possible in retrospect, or is hscript a dependency for this lib now?

**Comment by Gama11** [comment]
It's not possible unless we remove the `<haxelib name="hscript" />` from the `include.xml`, so the user would have to add that manually to his `Project.xml`, since there's no way to tell openfl to "include this lib if it's there". If a `haxelib` isn't there, it errors.

**Comment by Gama11** [comment]
We could try using `Compiler.addClassPath()` and check the output of `haxelib path hscript` in the initialization macro. Unfortunately there's no `Compiler.addHaxelib()` (https://github.com/HaxeFoundation/haxe/issues/4678).
Not sure if that would confuse OpenFL in some way.

**Pull request merged**